GetRefreshToken.php

```diff
@@ -1,43 +1,119 @@
+ * to generate an offline refresh token for installed application flows.
+ *
+ * <p>You can then use this refresh token to generate access tokens to
+ * authenticate against the ads API(s) you're using.
+ */
+class GetRefreshToken {
 
-	printf("Log in to your DFP account and open the following URL:\n%s\n\n", $authorizationUrl);
-	print "After approving the token enter the authorization code here: ";
-	$stdin = fopen('php://stdin', 'r');
-	$code = trim(fgets($stdin));
-	fclose($stdin);
-	print "\n";
+	/**
+	 * @var string the OAuth2 scope for the AdWords API
+	 * @see https://developers.google.com/adwords/api/docs/guides/authentication#scope
+	 */
+	const ADWORDS_API_SCOPE = 'https://www.googleapis.com/auth/adwords';
 
-	$user->SetOAuth2Info($OAuth2Handler->GetAccessToken($user->GetOAuth2Info(), $code, $redirectUri));
+	/**
+	 * @var string the OAuth2 scope for the DFP API
+	 * @see https://developers.google.com/doubleclick-publishers/docs/authentication#scope
+	 */
+	const DFP_API_SCOPE = 'https://www.googleapis.com/auth/dfp';
 
-	return $user->GetOAuth2Info();
-}
+	/**
+	 * @var string the Google OAuth2 authorization URI for OAuth2 requests
+	 * @see https://developers.google.com/identity/protocols/OAuth2InstalledApp#formingtheurl
+	 */
+	const AUTHORIZATION_URI = 'https://accounts.google.com/o/oauth2/v2/auth';
 
-if (__FILE__ != realpath($_SERVER['PHP_SELF'])) {
-	return;
-}
+	/**
+	 * @var string the redirect URI for OAuth2 installed application flows
+	 * @see https://developers.google.com/identity/protocols/OAuth2InstalledApp#formingtheurl
+	 */
+	const REDIRECT_URI = 'urn:ietf:wg:oauth:2.0:oob';
+
+	public static function main() {
+		$PRODUCTS = [
+			['AdWords', self::ADWORDS_API_SCOPE],
+			['DFP', self::DFP_API_SCOPE],
+			['AdWords and DFP', self::ADWORDS_API_SCOPE . ' ' . self::DFP_API_SCOPE]
+		];
+
+		$stdin = fopen('php://stdin', 'r');
+
+		print 'Enter your OAuth2 client ID here: ';
+		$clientId = trim(fgets($stdin));
+
+		print 'Enter your OAuth2 client secret here: ';
+		$clientSecret = trim(fgets($stdin));
+
+		print "Select the ads API you're using: [0] AdWords [1] DFP [2] Both\n";
+		$api = trim(fgets($stdin));
 
-try {
-	$user = new DfpUser(__DIR__ . '/config/auth.ini');
-	$user->LogAll();
-
-	$oauth2Info = GetOAuth2Credential($user);
-
-	printf("Your refresh token is: %s\n\n", $oauth2Info['refresh_token']);
-	printf("In your auth.ini file, edit the refresh_token line to be:\n"
-		. "refresh_token = \"%s\"\n", $oauth2Info['refresh_token']);
-} catch (OAuth2Exception $e) {
-	printf("%s\n", $e->getMessage());
-} catch (ValidationException $e) {
-	printf("%s\n", $e->getMessage());
-} catch (Exception $e) {
-	printf("An error has occurred: %s\n", $e->getMessage());
+		while (!is_numeric($api)
+			|| !(strval(intval($api)) === $api)
+			|| !(intval($api) >= 0 && intval($api) <= 2)) {
+			print "Please enter a valid number for the ads API you're using: [0] "
+				. "AdWords [1] DFP [2] Both\n";
+			$api = trim(fgets($stdin));
```

Comment:
nitpick: according to our guidelines, the `||` and `&&` should be at the end of line